File: 420-strong-password-checker/strong-password-checker.py
class Solution:
    def strongPasswordChecker(self, password: str) -> int:
        n = len(password)
        
        has_lower = any(c.islower() for c in password)
        has_upper = any(c.isupper() for c in password)
        has_digit = any(c.isdigit() for c in password)
        missing = (not has_lower) + (not has_upper) + (not has_digit)
        
        change = 0
        one = 0
        two = 0
        p = 2
        while p < n:
            if password[p] == password[p - 1] == password[p - 2]:
                length = 2
                while p < n and password[p] == password[p - 1]:
                    length += 1
                    p += 1
                change += length // 3
                if length % 3 == 0:
                    one += 1
                elif length % 3 == 1:
                    two += 1
            else:
                p += 1
        
        if n < 6:
            return max(missing, 6 - n)
        elif n <= 20:
            return max(missing, change)
        else:
            delete = n - 20
            change -= min(delete, one)
            change -= min(max(delete - one, 0), two * 2) // 2
            change -= max(delete - one - 2 * two, 0) // 3
            return delete + max(missing, change)

File: 420-strong-password-checker/strong-password-checker.cpp
class Solution {
public:
    int strongPasswordChecker(string password) {
        int n = password.size();
        bool hasLower = false, hasUpper = false, hasDigit = false;
        for (char c : password) {
            if (islower(c)) hasLower = true;
            if (isupper(c)) hasUpper = true;
            if (isdigit(c)) hasDigit = true;
        }
        int missing = (!hasLower) + (!hasUpper) + (!hasDigit);
        
        int change = 0, one = 0, two = 0;
        int p = 2;
        while (p < n) {
            if (password[p] == password[p-1] && password[p-1] == password[p-2]) {
                int length = 2;
                while (p < n && password[p] == password[p-1]) {
                    length++;
                    p++;
                }
                change += length / 3;
                if (length % 3 == 0) one++;
                else if (length % 3 == 1) two++;
            } else {
                p++;
            }
        }
        
        if (n < 6) {
            return max(missing, 6 - n);
        } else if (n <= 20) {
            return max(missing, change);
        } else {
            int del = n - 20;
            change -= min(del, one);
            change -= min(max(del - one, 0), two * 2) / 2;
            change -= max(del - one - 2 * two, 0) / 3;
            return del + max(missing, change);
        }
    }
};